library/std/src/sys/dbg.rs

```diff
@@ -0,0 +1,186 @@
+        debug_assert_eq!(info_size, KINFO_PROC_SIZE);
+
+        let mut reader = Cursor::new(kinfo_proc);
+        reader.seek(SeekFrom::Start(KINFO_PROC_FLAGS_OFFSET)).ok()?;
+        reader.read_exact(&mut flags).ok()?;
+        // Just in case, not limiting this to the little-endian systems.
+        let flags = i32::from_ne_bytes(flags);
+
+        if flags & P_TRACED != 0 {
+            Some(DebuggerPresence::Detected)
+        } else {
+            Some(DebuggerPresence::NotDetected)
+        }
+    }
+}
+
+#[cfg(target_os = "linux")]
+mod os {
+    use super::DebuggerPresence;
+    use crate::fs::File;
+    use crate::io::{BufRead, BufReader};
+
+    pub(super) fn is_debugger_present() -> Option<DebuggerPresence> {
+        let file = File::open("/proc/self/status").ok()?;
+        let reader = BufReader::new(file);
+
+        for line in reader.lines().flatten() {
+            if line.starts_with("TracerPid:") {
+                return if line.split(':').nth(1).map_or(false, |pid| pid.trim() != "0") {
+                    Some(DebuggerPresence::Detected)
+                } else {
+                    Some(DebuggerPresence::NotDetected)
+                };
+            }
+        }
+
+        None
+    }
+}
+
+// 32-bit macOS Rust targets are no longer shipped.
+#[cfg(not(any(
+    target_os = "windows",
+    all(target_os = "macos", target_pointer_width = "64"),
+    target_os = "freebsd",
+    target_os = "linux"
+)))]
+mod os {
+    pub(super) fn is_debugger_present() -> Option<super::DebuggerPresence> {
+        None
+    }
+}
+
+/// Detect the debugger presence.
+///
+/// The code does not try to detect the debugger at all costs (e.g., when anti-debugger
+/// tricks are at play), it relies on the interfaces provided by the OS.
+///
+/// Return value:
+/// * `None`: it's not possible to conclude whether the debugger is attached to this
+///    process or not. When checking for the presence of the debugger, the detection logic
+///    encountered an issue, such as the OS API throwing an error or the feature not being
+///    implemented.
+/// * `Some(DebuggerPresence::Detected)`: yes, the debugger is attached
+///   to this process.
+/// * `Some(DebuggerPresence::NotDetected)`: no, the debugger is not
+///    attached to this process.
+pub(crate) fn is_debugger_present() -> Option<DebuggerPresence> {
+    if cfg!(miri) { None } else { os::is_debugger_present() }
+}
+
+/// Execute the breakpoint instruction if the debugger presence is detected.
+/// Useful for breaking into the debugger without the need to set a breakpoint
+/// in the debugger.
+///
+/// Note that there is a race between attaching or detaching the debugger, and running the
+/// breakpoint instruction. In case the debugger is attached and the function is about
+/// to run the breakpoint instruction yet right before that the debugger detaches, the
+/// process will crash due to running the breakpoint instruction and the debugger not
+/// handling the trap exception.
```

Comment:
When we say there is a "race", usually "data race" is implied, so we should explain this is actually safe, because program termination is safe:

```suggestion
/// handling the trap exception. This is nonetheless memory-safe, like [`std::process::abort`] is.
```

Reply:
I see now, appreciate the suggested change! Let me fix the loose language there.